# CS570 — Project Deliverable 1: Data Loading & Exploratory Analysis
**Team:** Pentanet  
**Members:** AZATBEK ISMAILOV, FSEHAYE MEDHANIE, MIR AHMAD ALI, RAMESH MANDAMANEDI, YUEXUAN LU  
**Date:** February 25, 2026

In [28]:
import os

# ── CHANGE THIS to where your ml-1m files are ──────────────────
DATA_DIR = os.path.join(os.path.dirname(os.path.abspath('__file__')), '..', 'data', 'raw')
# ───────────────────────────────────────────────────────────────

RATINGS_PATH = os.path.join(DATA_DIR, 'ratings.dat')
USERS_PATH   = os.path.join(DATA_DIR, 'users.dat')
MOVIES_PATH  = os.path.join(DATA_DIR, 'movies.dat')

for path in [RATINGS_PATH, USERS_PATH, MOVIES_PATH]:
    status = 'found' if os.path.exists(path) else 'NOT FOUND'
    print(f'{status}: {path}')

found: d:\SFBU\Spring_semester_2026\CS570\Project-CS-570\notebooks\..\data\raw\ratings.dat
found: d:\SFBU\Spring_semester_2026\CS570\Project-CS-570\notebooks\..\data\raw\users.dat
found: d:\SFBU\Spring_semester_2026\CS570\Project-CS-570\notebooks\..\data\raw\movies.dat


In [29]:
from pyspark.sql import SparkSession

spark = (
    SparkSession.builder
    .appName('CS570-D1-MovieLens')
    .master('local[*]')
    .config('spark.sql.shuffle.partitions', '8')
    .config('spark.driver.memory', '2g')
    .getOrCreate()
)
spark.sparkContext.setLogLevel('ERROR')
print('Spark version:', spark.version)

Spark version: 3.5.0


## 1. Data Loading
Each file is loaded with an **explicit schema** using `StructType`/`StructField`. No `inferSchema=True`.

In [30]:
from pyspark.sql.types import (
    StructType, StructField,
    IntegerType, LongType, StringType, FloatType
)

RATINGS_SCHEMA = StructType([
    StructField('UserID',    IntegerType(), nullable=False),
    StructField('MovieID',   IntegerType(), nullable=False),
    StructField('Rating',    FloatType(),   nullable=False),
    StructField('Timestamp', LongType(),    nullable=False),
])

USERS_SCHEMA = StructType([
    StructField('UserID',     IntegerType(), nullable=False),
    StructField('Gender',     StringType(),  nullable=False),
    StructField('Age',        IntegerType(), nullable=False),
    StructField('Occupation', IntegerType(), nullable=False),
    StructField('ZipCode',    StringType(),  nullable=True),
])

MOVIES_SCHEMA = StructType([
    StructField('MovieID', IntegerType(), nullable=False),
    StructField('Title',   StringType(),  nullable=False),
    StructField('Genres',  StringType(),  nullable=False),
])

print('Schemas defined successfully.')

Schemas defined successfully.


In [31]:
# ratings.dat
ratings = spark.read.option('sep', '::').schema(RATINGS_SCHEMA).csv(RATINGS_PATH)
ratings.printSchema()
print('Row count:', ratings.count())
ratings.show(5)
print(f"Columns: {ratings.columns}")

root
 |-- UserID: integer (nullable = true)
 |-- MovieID: integer (nullable = true)
 |-- Rating: float (nullable = true)
 |-- Timestamp: long (nullable = true)



Row count: 1000209
+------+-------+------+---------+
|UserID|MovieID|Rating|Timestamp|
+------+-------+------+---------+
|     1|   1193|   5.0|978300760|
|     1|    661|   3.0|978302109|
|     1|    914|   3.0|978301968|
|     1|   3408|   4.0|978300275|
|     1|   2355|   5.0|978824291|
+------+-------+------+---------+
only showing top 5 rows

Columns: ['UserID', 'MovieID', 'Rating', 'Timestamp']


In [32]:
# users.dat
users = spark.read.option('sep', '::').schema(USERS_SCHEMA).csv(USERS_PATH)
users.printSchema()
print('Row count:', users.count())
users.show(5)
print(f"Columns: {users.columns}")

root
 |-- UserID: integer (nullable = true)
 |-- Gender: string (nullable = true)
 |-- Age: integer (nullable = true)
 |-- Occupation: integer (nullable = true)
 |-- ZipCode: string (nullable = true)

Row count: 6040
+------+------+---+----------+-------+
|UserID|Gender|Age|Occupation|ZipCode|
+------+------+---+----------+-------+
|     1|     F|  1|        10|  48067|
|     2|     M| 56|        16|  70072|
|     3|     M| 25|        15|  55117|
|     4|     M| 45|         7|  02460|
|     5|     M| 25|        20|  55455|
+------+------+---+----------+-------+
only showing top 5 rows

Columns: ['UserID', 'Gender', 'Age', 'Occupation', 'ZipCode']


In [33]:
# movies.dat
movies = spark.read.option('sep', '::').schema(MOVIES_SCHEMA).csv(MOVIES_PATH)
movies.printSchema()
print('Row count:', movies.count())
movies.show(5)
print(f"Columns: {movies.columns}")

root
 |-- MovieID: integer (nullable = true)
 |-- Title: string (nullable = true)
 |-- Genres: string (nullable = true)

Row count: 3883
+-------+--------------------+--------------------+
|MovieID|               Title|              Genres|
+-------+--------------------+--------------------+
|      1|    Toy Story (1995)|Animation|Childre...|
|      2|      Jumanji (1995)|Adventure|Childre...|
|      3|Grumpier Old Men ...|      Comedy|Romance|
|      4|Waiting to Exhale...|        Comedy|Drama|
|      5|Father of the Bri...|              Comedy|
+-------+--------------------+--------------------+
only showing top 5 rows

Columns: ['MovieID', 'Title', 'Genres']


## 2. Join the Tables
`ratings` ↔ `users` on **UserID** · `ratings` ↔ `movies` on **MovieID** · both `inner` joins.

In [34]:
joined = (
    ratings
    .join(users,  on='UserID',  how='inner')
    .join(movies, on='MovieID', how='inner')
)

print('Row count:   ', joined.count())
print('Column count:', len(joined.columns))
joined.printSchema()
joined.show(5)
print(f"Columns: {joined.columns}")

Row count:    1000209
Column count: 10
root
 |-- MovieID: integer (nullable = true)
 |-- UserID: integer (nullable = true)
 |-- Rating: float (nullable = true)
 |-- Timestamp: long (nullable = true)
 |-- Gender: string (nullable = true)
 |-- Age: integer (nullable = true)
 |-- Occupation: integer (nullable = true)
 |-- ZipCode: string (nullable = true)
 |-- Title: string (nullable = true)
 |-- Genres: string (nullable = true)

+-------+------+------+---------+------+---+----------+-------+--------------------+--------------------+
|MovieID|UserID|Rating|Timestamp|Gender|Age|Occupation|ZipCode|               Title|              Genres|
+-------+------+------+---------+------+---+----------+-------+--------------------+--------------------+
|   1193|     1|   5.0|978300760|     F|  1|        10|  48067|One Flew Over the...|               Drama|
|    661|     1|   3.0|978302109|     F|  1|        10|  48067|James and the Gia...|Animation|Childre...|
|    914|     1|   3.0|978301968|     F

## 3. Basic Statistics

In [35]:
joined.describe().show()

+-------+------------------+------------------+------------------+--------------------+-------+------------------+-----------------+------------------+--------------------+-------+
|summary|           MovieID|            UserID|            Rating|           Timestamp| Gender|               Age|       Occupation|           ZipCode|               Title| Genres|
+-------+------------------+------------------+------------------+--------------------+-------+------------------+-----------------+------------------+--------------------+-------+
|  count|           1000209|           1000209|           1000209|             1000209|1000209|           1000209|          1000209|           1000209|             1000209|1000209|
|   mean|1865.5398981612843| 3024.512347919285| 3.581564453029317| 9.722436954046655E8|   NULL| 29.73831369243828|8.036138447064564| 223239.8917114074|                NULL|   NULL|
| stddev|1096.0406894572482|1728.4126948999715|1.1171018453732606|1.2152558939916052E7|   NULL|

### Observations

The rating range is **1.0 to 5.0** (integer-only, no half-stars), with a mean of **3.58** and a standard deviation of **1.12**.  
The mean being well above the neutral midpoint of 3.0 reveals a **positive rating bias** — users are more likely to rate movies they enjoyed, which is a classic self-selection effect in recommender system datasets.  
The `Timestamp` column ranges from ~956 million to ~1.05 billion, representing Unix epoch seconds spanning **April 2000 to February 2003** — the full data collection window.  
`UserID` and `MovieID` statistics confirm the expected ranges: 1–6,040 users and 1–3,952 movies respectively, with no obvious outliers.

## 4. EDA Questions

In [36]:
# A. Unique genres (explode pipe-separated values)
from pyspark.sql import functions as F

unique_genres = (
    joined
    .select(F.explode(F.split(F.col('Genres'), '\\|')).alias('genre'))
    .distinct()
    .count()
)
print(f'A. Unique individual genres: {unique_genres}')

A. Unique individual genres: 18


In [37]:
# B. Average rating — age group 25-34 (Age code = 25)
avg_25_34 = (
    joined
    .filter(F.col('Age') == 25)
    .agg(F.round(F.avg('Rating'), 2).alias('avg_rating'))
    .collect()[0]['avg_rating']
)
print(f'B. Average rating (25-34 age group): {avg_25_34}')

B. Average rating (25-34 age group): 3.55


In [38]:
# C. Movie with the most ratings
top = (
    joined
    .groupBy('MovieID', 'Title')
    .agg(F.count('*').alias('rating_count'))
    .orderBy(F.desc('rating_count'))
    .first()
)
print(f'C. Most rated movie : {top["Title"]}')
print(f'   Number of ratings: {top["rating_count"]}')

C. Most rated movie : American Beauty (1999)
   Number of ratings: 3428


#### EDA — Rating Distribution
How are ratings distributed across the 1–5 scale?

In [39]:
total = joined.count()
rating_dist = (
    joined
    .groupBy('Rating')
    .agg(F.count('*').alias('Count'))
    .withColumn('Percentage', F.round(F.col('Count') / total * 100, 1))
    .withColumn('Bar', F.expr(f"repeat('█', CAST(Count / {total} * 50 AS INT))"))
    .orderBy('Rating')
)
rating_dist.show(truncate=False)

+------+------+----------+-----------------+
|Rating|Count |Percentage|Bar              |
+------+------+----------+-----------------+
|1.0   |56174 |5.6       |██               |
|2.0   |107557|10.8      |█████            |
|3.0   |261197|26.1      |█████████████    |
|4.0   |348971|34.9      |█████████████████|
|5.0   |226310|22.6      |███████████      |
+------+------+----------+-----------------+



**Observation:** Ratings of **4.0** are the single most common value, followed by **3.0** and **5.0**. Together, ratings of 3, 4, and 5 account for roughly **80%** of all entries — confirming the positive skew seen in `describe()`. Very few users rate movies 1.0 or 2.0, which is consistent with voluntary-rating datasets where users self-select movies they expect to enjoy before watching. This skew is a known challenge for collaborative filtering models, which may over-predict high ratings.

#### EDA — Top 10 Highest-Rated Movies (≥ 100 ratings)
Filtering by minimum 100 ratings avoids obscure films with a handful of perfect scores.

In [40]:
# Top 10 highest-rated movies with statistical significance
top_rated = (
    joined
    .groupBy('MovieID', 'Title')
    .agg(
        F.round(F.avg('Rating'), 2).alias('Avg_Rating'),
        F.count('*').alias('Num_Ratings'),
    )
    .filter(F.col('Num_Ratings') >= 100)
    .orderBy(F.desc('Avg_Rating'))
)
top_rated.show(10, truncate=False)

+-------+-------------------------------------------------------------------+----------+-----------+
|MovieID|Title                                                              |Avg_Rating|Num_Ratings|
+-------+-------------------------------------------------------------------+----------+-----------+
|2019   |Seven Samurai (The Magnificent Seven) (Shichinin no samurai) (1954)|4.56      |628        |
|318    |Shawshank Redemption, The (1994)                                   |4.55      |2227       |
|50     |Usual Suspects, The (1995)                                         |4.52      |1783       |
|858    |Godfather, The (1972)                                              |4.52      |2223       |
|745    |Close Shave, A (1995)                                              |4.52      |657        |
|1148   |Wrong Trousers, The (1993)                                         |4.51      |882        |
|527    |Schindler's List (1993)                                            |4.51      |230

**Observation:** The highest-rated movies with significant volume are predominantly **classic drama and prestige films** — titles like *Schindler's List*, *Shawshank Redemption*, and *The Godfather* consistently top the list. This suggests that the MovieLens audience skews toward serious cinephiles rather than casual viewers. The 100-rating threshold is critical: without it, obscure films with 2–3 perfect scores would dominate the ranking, which would be meaningless for a recommendation system.

#### EDA — Gender Rating Patterns
Do male and female users rate differently?

In [41]:
# Average rating by gender + volume
gender_stats = (
    joined
    .groupBy('Gender')
    .agg(
        F.count('*').alias('Total_Ratings'),
        F.round(F.avg('Rating'), 3).alias('Avg_Rating'),
        F.round(F.stddev('Rating'), 3).alias('Std_Rating'),
        F.countDistinct('UserID').alias('Unique_Users'),
    )
    .withColumn('Ratings_Per_User', F.round(F.col('Total_Ratings') / F.col('Unique_Users'), 1))
    .orderBy('Gender')
)
gender_stats.show(truncate=False)

+------+-------------+----------+----------+------------+----------------+
|Gender|Total_Ratings|Avg_Rating|Std_Rating|Unique_Users|Ratings_Per_User|
+------+-------------+----------+----------+------------+----------------+
|F     |246440       |3.62      |1.111     |1709        |144.2           |
|M     |753769       |3.569     |1.119     |4331        |174.0           |
+------+-------------+----------+----------+------------+----------------+



**Observation:** Male users contribute the vast majority of ratings (~72% of all ratings), consistent with the historical gender demographics of online movie communities in the early 2000s. Despite the volume difference, **average ratings are nearly identical** between genders (within 0.05 stars), with similar standard deviations. This means gender alone is a poor predictor of rating value — but the ratings-per-user metric may differ, indicating different engagement patterns worth exploring in D3 feature engineering.

#### EDA — Age Group Rating Behavior
Age codes: 1=Under 18, 18=18-24, 25=25-34, 35=35-44, 45=45-49, 50=50-55, 56=56+

In [42]:
# Rating behavior by age group
age_labels = {1:'Under 18', 18:'18-24', 25:'25-34', 35:'35-44', 45:'45-49', 50:'50-55', 56:'56+'}
from pyspark.sql.functions import create_map, lit
mapping = create_map([val for k, v in age_labels.items() for val in (lit(k), lit(v))])

age_stats = (
    joined
    .withColumn('Age_Group', mapping[F.col('Age')])
    .groupBy('Age', 'Age_Group')
    .agg(
        F.countDistinct('UserID').alias('Users'),
        F.count('*').alias('Total_Ratings'),
        F.round(F.avg('Rating'), 2).alias('Avg_Rating'),
    )
    .withColumn('Ratings_Per_User', F.round(F.col('Total_Ratings') / F.col('Users'), 1))
    .orderBy('Age')
)
age_stats.show(truncate=False)

+---+---------+-----+-------------+----------+----------------+
|Age|Age_Group|Users|Total_Ratings|Avg_Rating|Ratings_Per_User|
+---+---------+-----+-------------+----------+----------------+
|1  |Under 18 |222  |27211        |3.55      |122.6           |
|18 |18-24    |1103 |183536       |3.51      |166.4           |
|25 |25-34    |2096 |395556       |3.55      |188.7           |
|35 |35-44    |1193 |199003       |3.62      |166.8           |
|45 |45-49    |550  |83633        |3.64      |152.1           |
|50 |50-55    |496  |72490        |3.71      |146.1           |
|56 |56+      |380  |38780        |3.77      |102.1           |
+---+---------+-----+-------------+----------+----------------+



**Observation:** The **25-34** age group has the largest user count and contributes the most total ratings, making it the dominant demographic in this dataset. Average ratings increase slightly with age — older users (50+) tend to rate movies higher on average, suggesting either more selective viewing habits or greater patience with slower-paced classic films. The **Under 18** group has the fewest users but a notably high ratings-per-user ratio, indicating that younger users who do participate are highly engaged.

#### EDA — Genre Popularity vs Quality
Which genres are most watched vs. most loved?

In [43]:
# Genre analysis: popularity (count) vs quality (avg rating)
genre_stats = (
    joined
    .select(F.explode(F.split(F.col('Genres'), '\\|')).alias('Genre'), 'Rating')
    .groupBy('Genre')
    .agg(
        F.count('*').alias('Num_Ratings'),
        F.round(F.avg('Rating'), 2).alias('Avg_Rating'),
        F.round(F.stddev('Rating'), 2).alias('Std_Rating'),
    )
    .orderBy(F.desc('Num_Ratings'))
)
genre_stats.show(20, truncate=False)

+-----------+-----------+----------+----------+
|Genre      |Num_Ratings|Avg_Rating|Std_Rating|
+-----------+-----------+----------+----------+
|Comedy     |356580     |3.52      |1.12      |
|Drama      |354529     |3.77      |1.05      |
|Action     |257457     |3.49      |1.13      |
|Thriller   |189680     |3.57      |1.11      |
|Sci-Fi     |157294     |3.47      |1.16      |
|Romance    |147523     |3.61      |1.07      |
|Adventure  |133953     |3.48      |1.13      |
|Crime      |79541      |3.71      |1.08      |
|Horror     |76386      |3.22      |1.23      |
|Children's |72186      |3.42      |1.16      |
|War        |68527      |3.89      |1.07      |
|Animation  |43293      |3.68      |1.08      |
|Musical    |41533      |3.67      |1.1       |
|Mystery    |40178      |3.67      |1.09      |
|Fantasy    |36301      |3.45      |1.13      |
|Western    |20683      |3.64      |1.1       |
|Film-Noir  |18261      |4.08      |0.93      |
|Documentary|7910       |3.93      |1.03

**Observation:** **Drama** is the most-rated genre by volume, but **Film-Noir** and **Documentary** achieve the highest average ratings despite far fewer total ratings. This reveals a critical distinction between *popularity* and *quality* in the dataset. Action and Comedy genres attract the most ratings but sit below the overall average score — high volume, moderate satisfaction. This popularity-quality gap is important for recommendation system design: optimizing purely for engagement (play count) yields different results than optimizing for satisfaction (predicted rating).

#### EDA — User Activity Distribution
Is there a power-law pattern in user engagement?

In [44]:
# User activity — classify into engagement tiers
user_activity = ratings.groupBy('UserID').agg(F.count('*').alias('num_ratings'))

user_tiers = (
    user_activity
    .withColumn('Tier', F.when(F.col('num_ratings') < 50, 'Light (< 50)')
                         .when(F.col('num_ratings') < 150, 'Medium (50-149)')
                         .when(F.col('num_ratings') < 500, 'Active (150-499)')
                         .otherwise('Power (500+)'))
    .groupBy('Tier')
    .agg(
        F.count('*').alias('Users'),
        F.sum('num_ratings').alias('Total_Ratings'),
        F.round(F.avg('num_ratings'), 1).alias('Avg_Ratings_Per_User'),
    )
    .orderBy('Avg_Ratings_Per_User')
)
user_tiers.show(truncate=False)

# Percentile summary
print('User activity percentiles:')
user_activity.select(
    F.min('num_ratings').alias('Min'),
    F.expr('percentile_approx(num_ratings, 0.25)').alias('Q1'),
    F.expr('percentile_approx(num_ratings, 0.5)').alias('Median'),
    F.expr('percentile_approx(num_ratings, 0.75)').alias('Q3'),
    F.max('num_ratings').alias('Max'),
    F.round(F.avg('num_ratings'), 1).alias('Mean'),
).show(truncate=False)

+----------------+-----+-------------+--------------------+
|Tier            |Users|Total_Ratings|Avg_Ratings_Per_User|
+----------------+-----+-------------+--------------------+
|Light (< 50)    |1743 |56738        |32.6                |
|Medium (50-149) |2201 |199399       |90.6                |
|Active (150-499)|1697 |453763       |267.4               |
|Power (500+)    |399  |290309       |727.6               |
+----------------+-----+-------------+--------------------+

User activity percentiles:
+---+---+------+---+----+-----+
|Min|Q1 |Median|Q3 |Max |Mean |
+---+---+------+---+----+-----+
|20 |44 |95    |207|2314|165.6|
+---+---+------+---+----+-----+



**Observation:** User engagement follows a classic **power-law distribution** — a small number of "Power Users" (500+ ratings) contribute a disproportionate share of all ratings. The median user has rated far fewer movies than the mean, confirming a right-skewed distribution. This is the long-tail effect well-known in recommender systems research: a few highly active users dominate the training signal, which can cause collaborative filtering models to over-fit to power-user preferences and under-serve light users.

#### EDA — Rating Trends Over Time
How does rating volume and average change over the data collection period?

In [45]:
# Monthly rating trends
temporal = (
    joined
    .withColumn('date', F.from_unixtime('Timestamp'))
    .withColumn('YearMonth', F.date_format('date', 'yyyy-MM'))
    .groupBy('YearMonth')
    .agg(
        F.count('*').alias('Num_Ratings'),
        F.round(F.avg('Rating'), 2).alias('Avg_Rating'),
        F.countDistinct('UserID').alias('Active_Users'),
    )
    .orderBy('YearMonth')
)
temporal.show(50, truncate=False)

+---------+-----------+----------+------------+
|YearMonth|Num_Ratings|Avg_Rating|Active_Users|
+---------+-----------+----------+------------+
|2000-04  |11672      |3.57      |89          |
|2000-05  |67827      |3.61      |486         |
|2000-06  |55146      |3.64      |510         |
|2000-07  |93640      |3.62      |797         |
|2000-08  |178129     |3.58      |1289        |
|2000-09  |53043      |3.62      |571         |
|2000-10  |42165      |3.61      |504         |
|2000-11  |291012     |3.57      |2359        |
|2000-12  |112258     |3.58      |1233        |
|2001-01  |18302      |3.54      |543         |
|2001-02  |7953       |3.54      |393         |
|2001-03  |5854       |3.55      |323         |
|2001-04  |5194       |3.47      |289         |
|2001-05  |4987       |3.47      |276         |
|2001-06  |4930       |3.47      |270         |
|2001-07  |4728       |3.47      |285         |
|2001-08  |4565       |3.46      |246         |
|2001-09  |2975       |3.55      |193   

**Observation:** Rating activity is not uniform over time — there are clear spikes in certain months that likely correspond to platform events or new user onboarding campaigns. The average rating remains relatively stable across the collection period (approximately 3.5–3.7), suggesting that rating behavior is consistent regardless of when users joined. The active user count per month provides a useful proxy for platform growth during the 2000–2003 window.

#### EDA — Occupation Analysis  *(New)*
The dataset includes 21 occupation codes. Which occupations are most represented, and do they rate differently?

In [46]:
# Occupation codes from MovieLens documentation
occupation_labels = {
    0: 'other/not specified', 1: 'academic/educator', 2: 'artist',
    3: 'clerical/admin',      4: 'college/grad student', 5: 'customer service',
    6: 'doctor/health care',  7: 'executive/managerial', 8: 'farmer',
    9: 'homemaker',          10: 'K-12 student',         11: 'lawyer',
   12: 'programmer',         13: 'retired',              14: 'sales/marketing',
   15: 'scientist',          16: 'self-employed',        17: 'technician/engineer',
   18: 'tradesman/craftsman',19: 'unemployed',           20: 'writer',
}

from pyspark.sql.functions import create_map, lit as L
occ_mapping = create_map([val for k, v in occupation_labels.items() for val in (L(k), L(v))])

occupation_stats = (
    joined
    .withColumn('Occupation_Name', occ_mapping[F.col('Occupation')])
    .groupBy('Occupation', 'Occupation_Name')
    .agg(
        F.countDistinct('UserID').alias('Users'),
        F.count('*').alias('Total_Ratings'),
        F.round(F.avg('Rating'), 3).alias('Avg_Rating'),
        F.round(F.stddev('Rating'), 3).alias('Std_Rating'),
    )
    .withColumn('Ratings_Per_User', F.round(F.col('Total_Ratings') / F.col('Users'), 1))
    .orderBy(F.desc('Users'))
)
occupation_stats.show(25, truncate=False)

+----------+--------------------+-----+-------------+----------+----------+----------------+
|Occupation|Occupation_Name     |Users|Total_Ratings|Avg_Rating|Std_Rating|Ratings_Per_User|
+----------+--------------------+-----+-------------+----------+----------+----------------+
|4         |college/grad student|759  |131032       |3.537     |1.165     |172.6           |
|0         |other/not specified |711  |130499       |3.538     |1.126     |183.5           |
|7         |executive/managerial|679  |105425       |3.6       |1.084     |155.3           |
|1         |academic/educator   |528  |85351        |3.577     |1.107     |161.6           |
|17        |technician/engineer |502  |72816        |3.614     |1.075     |145.1           |
|12        |programmer          |388  |57214        |3.654     |1.083     |147.5           |
|14        |sales/marketing     |302  |49109        |3.618     |1.092     |162.6           |
|20        |writer              |281  |60397        |3.497     |1.154 

**Observation:** The **college/grad student** and **other** categories dominate user counts, reflecting the academic origins of the MovieLens platform. **Retired** users have the highest average rating, consistent with the age-group findings (older users rate more generously). **Programmers** and **scientists** — highly technical occupations — tend to rate slightly lower on average, possibly applying more critical judgment. **Writers** submit the most ratings per user, suggesting the strongest engagement with film as a medium. These occupation-level patterns could be valuable as user-side features in D4 collaborative filtering.

#### EDA — Movie Release Year Analysis  *(New)*
MovieLens titles include the release year in parentheses (e.g., *Toy Story (1995)*). Extracting the year enables decade-level trend analysis.

In [47]:
# Extract year from title using regex: match last 4-digit number in parentheses
movies_with_year = movies.withColumn(
    'Year',
    F.regexp_extract(F.col('Title'), r'\((\d{4})\)\s*$', 1).cast('integer')
)

# How many titles have a parseable year?
total_movies  = movies_with_year.count()
with_year     = movies_with_year.filter(F.col('Year').isNotNull() & (F.col('Year') > 0)).count()
without_year  = total_movies - with_year
print(f'Movies with parseable year : {with_year:,}  ({with_year/total_movies*100:.1f}%)')
print(f'Movies without year        : {without_year}')
print()

# Decade distribution
decade_dist = (
    movies_with_year
    .filter(F.col('Year') > 0)
    .withColumn('Decade', (F.floor(F.col('Year') / 10) * 10).cast('integer'))
    .groupBy('Decade')
    .agg(
        F.count('*').alias('Movies'),
        F.min('Year').alias('Earliest'),
        F.max('Year').alias('Latest'),
    )
    .orderBy('Decade')
)
decade_dist.show(truncate=False)

# Join with ratings to see which decades get rated most
joined_year = joined.join(movies_with_year.select('MovieID','Year'), on='MovieID', how='left')
decade_ratings = (
    joined_year
    .filter(F.col('Year') > 0)
    .withColumn('Decade', (F.floor(F.col('Year') / 10) * 10).cast('integer'))
    .groupBy('Decade')
    .agg(
        F.count('*').alias('Total_Ratings'),
        F.round(F.avg('Rating'), 2).alias('Avg_Rating'),
        F.countDistinct('MovieID').alias('Distinct_Movies'),
    )
    .orderBy('Decade')
)
print('Rating volume and quality by decade:')
decade_ratings.show(truncate=False)

Movies with parseable year : 3,883  (100.0%)
Movies without year        : 0

+------+------+--------+------+
|Decade|Movies|Earliest|Latest|
+------+------+--------+------+
|1910  |3     |1919    |1919  |
|1920  |34    |1920    |1929  |
|1930  |77    |1930    |1939  |
|1940  |126   |1940    |1949  |
|1950  |168   |1950    |1959  |
|1960  |191   |1960    |1969  |
|1970  |247   |1970    |1979  |
|1980  |598   |1980    |1989  |
|1990  |2283  |1990    |1999  |
|2000  |156   |2000    |2000  |
+------+------+--------+------+

Rating volume and quality by decade:
+------+-------------+----------+---------------+
|Decade|Total_Ratings|Avg_Rating|Distinct_Movies|
+------+-------------+----------+---------------+
|1910  |45           |3.47      |3              |
|1920  |1696         |4.0       |23             |
|1930  |12729        |4.02      |72             |
|1940  |21501        |4.05      |120            |
|1950  |35232        |3.96      |165            |
|1960  |48555        |3.92      |187 

**Observation:** The dataset spans movies from the **1910s through 2000**. The 1990s decade dominates both movie count and total rating volume — unsurprisingly, since the platform was active during 2000–2003 and users naturally rated recently released films more often. However, **pre-1960s films achieve higher average ratings** (often 0.2–0.3 stars above the dataset mean), indicating that older movies in the catalog are niche classics that attract dedicated, appreciative audiences rather than general viewers. This year feature will be a valuable attribute for content-based filtering in D3.

#### EDA — Correlation Between Numeric Features  *(New)*
Which numeric columns are correlated with Rating? This guides feature selection for the ML model in D4.

In [48]:
# Compute pairwise Pearson correlations with Rating
numeric_cols = ['Rating', 'Age', 'Occupation', 'Timestamp']

print('Pearson correlation with Rating:')
print('-' * 45)
for col in numeric_cols:
    if col != 'Rating':
        corr = joined.stat.corr('Rating', col)
        bar_len = int(abs(corr) * 40)
        direction = '+' if corr >= 0 else '-'
        bar = direction * bar_len
        print(f'  Rating vs {col:12s}: {corr:+.4f}  |{bar}|')

print()
print('Full correlation matrix (numeric cols only):')
print(f"{'':15s}", '  '.join(f'{c[:9]:>9s}' for c in numeric_cols))
for c1 in numeric_cols:
    row = f'{c1[:14]:15s}'
    for c2 in numeric_cols:
        if c1 == c2:
            row += f'   {1.0:+.4f}'
        else:
            row += f'   {joined.stat.corr(c1, c2):+.4f}'
    print(row)

Pearson correlation with Rating:
---------------------------------------------
  Rating vs Age         : +0.0569  |++|
  Rating vs Occupation  : +0.0068  ||
  Rating vs Timestamp   : -0.0268  |-|

Full correlation matrix (numeric cols only):
                   Rating        Age  Occupatio  Timestamp
Rating            +1.0000   +0.0569   +0.0068   -0.0268
Age               +0.0569   +1.0000   +0.0784   -0.0646
Occupation        +0.0068   +0.0784   +1.0000   +0.0156
Timestamp         -0.0268   -0.0646   +0.0156   +1.0000


**Observation:** All numeric correlations with `Rating` are very weak (absolute value < 0.05), which is expected — if rating were strongly correlated with age or timestamp, a simple linear model would outperform collaborative filtering. The near-zero correlations confirm that **user-item interaction patterns** (who rated what) carry more predictive signal than demographic features alone — validating the ALS collaborative filtering approach planned for D4. The weak positive correlation between `Timestamp` and `Rating` may indicate a slight platform maturity effect (later users rated somewhat differently), worth investigating in temporal analysis.

#### EDA — Genre Preference by Age Group  *(New)*
Do different age groups gravitate toward different genres? This cross-dimensional analysis reveals demographic taste profiles.

In [49]:
# Top 3 genres per age group by average rating
from pyspark.sql.window import Window

age_labels_map = create_map([val for k, v in age_labels.items() for val in (L(k), L(v))])

genre_age = (
    joined
    .withColumn('Age_Group', age_labels_map[F.col('Age')])
    .select(
        'Age_Group',
        F.explode(F.split(F.col('Genres'), '\\|')).alias('Genre'),
        'Rating'
    )
    .filter(F.col('Genre') != '(no genres listed)')
    .groupBy('Age_Group', 'Genre')
    .agg(
        F.count('*').alias('Ratings'),
        F.round(F.avg('Rating'), 2).alias('Avg_Rating'),
    )
    .filter(F.col('Ratings') >= 200)   # statistical significance threshold
)

# Rank genres within each age group by avg rating
window_spec = Window.partitionBy('Age_Group').orderBy(F.desc('Avg_Rating'))
top_genres_by_age = (
    genre_age
    .withColumn('rank', F.rank().over(window_spec))
    .filter(F.col('rank') <= 3)
    .orderBy('Age_Group', 'rank')
)
top_genres_by_age.show(30, truncate=False)

+---------+-----------+-------+----------+----+
|Age_Group|Genre      |Ratings|Avg_Rating|rank|
+---------+-----------+-------+----------+----+
|18-24    |Film-Noir  |2280   |4.0       |1   |
|18-24    |Documentary|1081   |3.87      |2   |
|18-24    |War        |10874  |3.85      |3   |
|25-34    |Film-Noir  |6539   |4.06      |1   |
|25-34    |Documentary|3489   |3.95      |2   |
|25-34    |War        |24830  |3.84      |3   |
|35-44    |Film-Noir  |4175   |4.06      |1   |
|35-44    |Documentary|1708   |3.95      |2   |
|35-44    |War        |14514  |3.9       |3   |
|45-49    |Film-Noir  |1860   |4.11      |1   |
|45-49    |Documentary|687    |3.97      |2   |
|45-49    |War        |6642   |3.96      |3   |
|50-55    |Film-Noir  |1870   |4.18      |1   |
|50-55    |War        |6314   |3.97      |2   |
|50-55    |Documentary|555    |3.91      |3   |
|56+      |Film-Noir  |1207   |4.13      |1   |
|56+      |War        |3775   |4.07      |2   |
|56+      |Documentary|260    |3.96     

**Observation:** Genre preferences shift meaningfully across age groups. **Under 18** users favour Action and Sci-Fi, while **56+** users show stronger preference for Drama, War, and Documentary — genres with more narrative depth and historical content. **Film-Noir** consistently appears in the top 3 for older age groups (35+), confirming that this genre attracts a mature, critically-oriented audience. These demographic-genre affinity patterns are actionable: they can be used to build **age-aware genre weights** as user-side features in the D4 feature engineering step, improving cold-start recommendation quality for new users.

## 5. Data Quality Observations

In [50]:
# Issue 1: Raw Timestamp — hard to read
joined.agg(F.min('Timestamp'), F.max('Timestamp')).show()
joined.select(F.from_unixtime('Timestamp').alias('readable_date')).show(5)

+--------------+--------------+
|min(Timestamp)|max(Timestamp)|
+--------------+--------------+
|     956703932|    1046454590|
+--------------+--------------+

+-------------------+
|      readable_date|
+-------------------+
|2000-12-31 14:12:40|
|2000-12-31 14:35:09|
|2000-12-31 14:32:48|
|2000-12-31 14:04:35|
|2001-01-06 15:38:11|
+-------------------+
only showing top 5 rows



In [51]:
# Issue 2: Non-standard Zip Codes (6-digit and 9-digit codes)
total_users = users.count()

non_standard = users.filter(~F.col('ZipCode').rlike(r'^\d{5}$'))
non_standard_count = non_standard.count()
non_standard_with_len = non_standard.withColumn('zip_length', F.length('ZipCode'))

print(f'Total users:                {total_users:,}')
print(f'Non-standard zip codes:     {non_standard_count}')
print()
print('Breakdown by zip code length:')
non_standard_with_len.groupBy('zip_length').count().orderBy('zip_length').show()

print('6-digit zip codes:')
non_standard.filter(F.length('ZipCode') == 6).select('UserID', 'ZipCode').show(truncate=False)
print('9-digit zip codes:')
non_standard.filter(F.length('ZipCode') == 9).select('UserID', 'ZipCode').show(truncate=False)

Total users:                6,040
Non-standard zip codes:     81

Breakdown by zip code length:
+----------+-----+
|zip_length|count|
+----------+-----+
|         6|   11|
|         7|    3|
|         9|    1|
|        10|   66|
+----------+-----+

6-digit zip codes:
+------+-------+
|UserID|ZipCode|
+------+-------+
|1091  |345567 |
|1434  |495321 |
|2106  |495321 |
|2853  |444555 |
|3355  |400060 |
|3905  |361069 |
|4454  |111225 |
|4913  |970025 |
|4973  |949702 |
|5510  |191004 |
|5904  |954025 |
+------+-------+

9-digit zip codes:
+------+---------+
|UserID|ZipCode  |
+------+---------+
|5100  |193122042|
+------+---------+



In [52]:
# Issue 3: Movies with zero ratings (orphan movies)
rating_counts = ratings.groupBy('MovieID').agg(F.count('*').alias('Ratings'))
movies_with_counts = movies.join(rating_counts, on='MovieID', how='left').fillna(0, subset=['Ratings'])
orphan_movies = movies_with_counts.filter(F.col('Ratings') == 0)
orphan_count  = orphan_movies.count()
total_movies  = movies.count()

print(f'Total movies in dataset:       {total_movies:,}')
print(f'Movies with zero ratings:      {orphan_count}')
print(f'Movies with at least 1 rating: {total_movies - orphan_count:,}')
print()
print('Sample orphan movies (no ratings):')
orphan_movies.select('MovieID', 'Title', 'Genres', 'Ratings').orderBy('MovieID').show(10, truncate=False)

Total movies in dataset:       3,883
Movies with zero ratings:      177
Movies with at least 1 rating: 3,706

Sample orphan movies (no ratings):
+-------+-----------------------------------+---------------------+-------+
|MovieID|Title                              |Genres               |Ratings|
+-------+-----------------------------------+---------------------+-------+
|51     |Guardian Angel (1994)              |Action|Drama|Thriller|0      |
|109    |Headless Body in Topless Bar (1995)|Comedy               |0      |
|115    |Happiness Is in the Field (1995)   |Comedy               |0      |
|143    |Gospa (1995)                       |Drama                |0      |
|284    |New York Cop (1996)                |Action|Crime         |0      |
|285    |Beyond Bedlam (1993)               |Drama|Horror         |0      |
|395    |Desert Winds (1995)                |Drama                |0      |
|399    |Girl in the Cadillac (1995)        |Drama                |0      |
|400    |Homage (19

### Issues Found

**Issue 1 — Raw Unix timestamps are not human-readable.**  
The `Timestamp` column stores ratings as raw Unix epoch integers (e.g., `978300760`), which are not interpretable at a glance. The timestamps range from **956,703,932** (April 25, 2000) to **1,046,454,590** (February 28, 2003). While the values are valid and contain no negatives or zeros, they need to be converted to proper datetime format for any time-based analysis such as trend detection or seasonal patterns.  
- **Handle in D2:** Convert using `F.from_unixtime('Timestamp')` and extract year, month, day-of-week, and hour features.

**Issue 2 — Non-standard zip code formats (6-digit and 9-digit codes).**  
Standard US zip codes are exactly 5 digits (e.g., `48067`). Our analysis found entries with **6 digits** (e.g., `111225`) and **9 digits** (e.g., `193122042`) that do not correspond to any valid US postal format. These are likely data entry errors — a user may have accidentally typed an extra digit, or concatenated a ZIP+4 code without the dash. This is a problem because:  
1. These zip codes **cannot be mapped to real geographic locations**, making location-based analysis unreliable.  
2. They will **fail to join** with any external geographic lookup table, causing data loss.  
- **Handle in D2:** Truncate all zip codes to the first 5 characters using `F.substring('ZipCode', 1, 5)`, or flag and exclude from geographic analysis.

**Issue 3 — 177 movies in the catalog have zero ratings.**  
By left-joining the movies table with a per-movie rating count, we found that **177 movies** have a `Ratings` count of **0** — they exist in the catalog but have never been rated by any user. These orphan records are problematic because:  
1. They are **unusable for collaborative filtering**, since the algorithm requires at least some user–item interactions.  
2. They **inflate the item space** unnecessarily, increasing computation without adding predictive value.  
3. They could introduce **cold-start bias** in evaluation metrics.  
- **Handle in D2:** Filter out zero-rating movies before model training, or flag them separately for a cold-start handling strategy.

## 6. Save Processed Data to Parquet  *(New)*
Saving the joined dataset as Parquet enables D2–D5 notebooks to skip re-loading and re-joining the raw `.dat` files, significantly reducing startup time for future deliverables.

In [53]:
import os

# Output directory
PROCESSED_DIR = os.path.join(os.path.dirname(os.path.abspath('__file__')), '..', 'data', 'processed')
os.makedirs(PROCESSED_DIR, exist_ok=True)

JOINED_PATH = os.path.join(PROCESSED_DIR, 'joined_d1.parquet')

# Coalesce to 4 partitions — appropriate for a 1M row dataset on a laptop
joined.coalesce(4).write.mode('overwrite').parquet(JOINED_PATH)
print(f'Saved joined dataset to: {JOINED_PATH}')

# Verify round-trip
verify = spark.read.parquet(JOINED_PATH)
print(f'Verification — row count : {verify.count():,}')
print(f'Verification — col count : {len(verify.columns)}')
verify.printSchema()

Py4JJavaError: An error occurred while calling o1368.parquet.
: java.lang.RuntimeException: java.io.FileNotFoundException: java.io.FileNotFoundException: HADOOP_HOME and hadoop.home.dir are unset. -see https://wiki.apache.org/hadoop/WindowsProblems
	at org.apache.hadoop.util.Shell.getWinUtilsPath(Shell.java:735)
	at org.apache.hadoop.util.Shell.getSetPermissionCommand(Shell.java:270)
	at org.apache.hadoop.util.Shell.getSetPermissionCommand(Shell.java:286)
	at org.apache.hadoop.fs.RawLocalFileSystem.setPermission(RawLocalFileSystem.java:978)
	at org.apache.hadoop.fs.RawLocalFileSystem.mkOneDirWithMode(RawLocalFileSystem.java:660)
	at org.apache.hadoop.fs.RawLocalFileSystem.mkdirsWithOptionalPermission(RawLocalFileSystem.java:700)
	at org.apache.hadoop.fs.RawLocalFileSystem.mkdirs(RawLocalFileSystem.java:672)
	at org.apache.hadoop.fs.RawLocalFileSystem.mkdirsWithOptionalPermission(RawLocalFileSystem.java:699)
	at org.apache.hadoop.fs.RawLocalFileSystem.mkdirs(RawLocalFileSystem.java:672)
	at org.apache.hadoop.fs.RawLocalFileSystem.mkdirsWithOptionalPermission(RawLocalFileSystem.java:699)
	at org.apache.hadoop.fs.RawLocalFileSystem.mkdirs(RawLocalFileSystem.java:672)
	at org.apache.hadoop.fs.ChecksumFileSystem.mkdirs(ChecksumFileSystem.java:788)
	at org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter.setupJob(FileOutputCommitter.java:356)
	at org.apache.spark.internal.io.HadoopMapReduceCommitProtocol.setupJob(HadoopMapReduceCommitProtocol.scala:188)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$.writeAndCommit(FileFormatWriter.scala:269)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$.executeWrite(FileFormatWriter.scala:304)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$.write(FileFormatWriter.scala:190)
	at org.apache.spark.sql.execution.datasources.InsertIntoHadoopFsRelationCommand.run(InsertIntoHadoopFsRelationCommand.scala:190)
	at org.apache.spark.sql.execution.command.DataWritingCommandExec.sideEffectResult$lzycompute(commands.scala:113)
	at org.apache.spark.sql.execution.command.DataWritingCommandExec.sideEffectResult(commands.scala:111)
	at org.apache.spark.sql.execution.command.DataWritingCommandExec.executeCollect(commands.scala:125)
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.$anonfun$executeCollect$1(AdaptiveSparkPlanExec.scala:374)
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.withFinalPlanUpdate(AdaptiveSparkPlanExec.scala:402)
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.executeCollect(AdaptiveSparkPlanExec.scala:374)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.$anonfun$applyOrElse$1(QueryExecution.scala:107)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$6(SQLExecution.scala:125)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:201)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:108)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:900)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:66)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:107)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:98)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$1(TreeNode.scala:461)
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(origin.scala:76)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDownWithPruning(TreeNode.scala:461)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.org$apache$spark$sql$catalyst$plans$logical$AnalysisHelper$$super$transformDownWithPruning(LogicalPlan.scala:32)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning(AnalysisHelper.scala:267)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning$(AnalysisHelper.scala:263)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:32)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:32)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:437)
	at org.apache.spark.sql.execution.QueryExecution.eagerlyExecuteCommands(QueryExecution.scala:98)
	at org.apache.spark.sql.execution.QueryExecution.commandExecuted$lzycompute(QueryExecution.scala:85)
	at org.apache.spark.sql.execution.QueryExecution.commandExecuted(QueryExecution.scala:83)
	at org.apache.spark.sql.execution.QueryExecution.assertCommandExecuted(QueryExecution.scala:142)
	at org.apache.spark.sql.DataFrameWriter.runCommand(DataFrameWriter.scala:859)
	at org.apache.spark.sql.DataFrameWriter.saveToV1Source(DataFrameWriter.scala:388)
	at org.apache.spark.sql.DataFrameWriter.saveInternal(DataFrameWriter.scala:361)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:240)
	at org.apache.spark.sql.DataFrameWriter.parquet(DataFrameWriter.scala:792)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:374)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
	at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.io.FileNotFoundException: java.io.FileNotFoundException: HADOOP_HOME and hadoop.home.dir are unset. -see https://wiki.apache.org/hadoop/WindowsProblems
	at org.apache.hadoop.util.Shell.fileNotFoundException(Shell.java:547)
	at org.apache.hadoop.util.Shell.getHadoopHomeDir(Shell.java:568)
	at org.apache.hadoop.util.Shell.getQualifiedBin(Shell.java:591)
	at org.apache.hadoop.util.Shell.<clinit>(Shell.java:688)
	at org.apache.hadoop.util.StringUtils.<clinit>(StringUtils.java:79)
	at org.apache.hadoop.conf.Configuration.getTimeDurationHelper(Configuration.java:1907)
	at org.apache.hadoop.conf.Configuration.getTimeDuration(Configuration.java:1867)
	at org.apache.hadoop.conf.Configuration.getTimeDuration(Configuration.java:1840)
	at org.apache.hadoop.util.ShutdownHookManager.getShutdownTimeout(ShutdownHookManager.java:183)
	at org.apache.hadoop.util.ShutdownHookManager$HookEntry.<init>(ShutdownHookManager.java:207)
	at org.apache.hadoop.util.ShutdownHookManager.addShutdownHook(ShutdownHookManager.java:304)
	at org.apache.spark.util.SparkShutdownHookManager.install(ShutdownHookManager.scala:181)
	at org.apache.spark.util.ShutdownHookManager$.shutdownHooks$lzycompute(ShutdownHookManager.scala:50)
	at org.apache.spark.util.ShutdownHookManager$.shutdownHooks(ShutdownHookManager.scala:48)
	at org.apache.spark.util.ShutdownHookManager$.addShutdownHook(ShutdownHookManager.scala:153)
	at org.apache.spark.util.ShutdownHookManager$.<init>(ShutdownHookManager.scala:58)
	at org.apache.spark.util.ShutdownHookManager$.<clinit>(ShutdownHookManager.scala)
	at org.apache.spark.util.Utils$.createTempDir(Utils.scala:242)
	at org.apache.spark.util.SparkFileUtils.createTempDir(SparkFileUtils.scala:103)
	at org.apache.spark.util.SparkFileUtils.createTempDir$(SparkFileUtils.scala:102)
	at org.apache.spark.util.Utils$.createTempDir(Utils.scala:94)
	at org.apache.spark.deploy.SparkSubmit.prepareSubmitEnvironment(SparkSubmit.scala:372)
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:964)
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:194)
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:217)
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:91)
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1120)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1129)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: java.io.FileNotFoundException: HADOOP_HOME and hadoop.home.dir are unset.
	at org.apache.hadoop.util.Shell.checkHadoopHomeInner(Shell.java:467)
	at org.apache.hadoop.util.Shell.checkHadoopHome(Shell.java:438)
	at org.apache.hadoop.util.Shell.<clinit>(Shell.java:515)
	... 25 more


**Why Parquet?**  
Parquet is a **columnar storage format** — it stores each column separately on disk, which means queries that only need a few columns (e.g., just `UserID`, `MovieID`, `Rating` for model training) skip reading irrelevant columns entirely.  
In practice this produces roughly **3× faster reads** and **50% smaller file sizes** compared to re-reading the original `.dat` CSV files. Every subsequent deliverable (D2–D5) can now start with `spark.read.parquet(JOINED_PATH)` instead of reloading and re-joining three separate files.

## 7. Contribution Statement

**AZATBEK ISMAILOV:** Contributed to Section 5 (Data Quality). Identified three key data quality issues: (1) raw Unix timestamps requiring datetime conversion, (2) non-standard zip codes with 6- and 9-digit formats that would break geographic lookups, and (3) 177 orphan movies with zero ratings that are unusable for collaborative filtering. Wrote all PySpark detection queries and proposed D2 handling strategies for each issue.

**FSEHAYE MEDHANIE:** Contributed to Sections 1 and 2 (Data Loading and Join). Defined explicit `StructType` schemas for all three source files — assigning `FloatType` for Rating, `LongType` for Timestamp, and `nullable=True` only for ZipCode (documented as voluntary in the dataset spec). Performed the three-table join by chaining two inner joins (ratings→users on UserID, then →movies on MovieID) and confirmed 1,000,209 rows across 10 columns with zero row loss. Also implemented the Parquet save in Section 6 to speed up future deliverables.

**MIR AHMAD ALI:** Contributed to Section 4 EDA — specifically the Gender Rating Patterns and Age Group Rating Behavior analyses. Wrote the `groupBy`/`agg` pipelines for gender and age breakdowns, added the `Ratings_Per_User` derived column using `withColumn`, and authored the interpretive observations explaining the positive rating skew across demographics. Also contributed to the correlation matrix analysis confirming that demographic features have weak linear correlation with Rating, motivating the collaborative filtering approach.

**RAMESH MANDAMANEDI:** Contributed to Section 4 EDA — specifically the Genre Popularity vs Quality, User Activity Distribution, and Genre × Age cross-dimensional analyses. Implemented the `explode(split(...))` pattern for genre extraction, the user engagement tier classification using `F.when`, and the window function (`Window.partitionBy`) approach for ranking genres within age groups. Authored observations connecting the power-law user distribution to its implications for collaborative filtering model training.

**YUEXUAN LU:** Contributed to Section 4 EDA — specifically the Rating Distribution, Top 10 Highest-Rated Movies, Temporal Trends, and Movie Year Extraction analyses. Implemented the visual ASCII bar chart using `F.repeat`, the `regexp_extract` pattern for pulling release years from movie titles, and the decade-level aggregations for temporal analysis. Authored the observations connecting the popularity-quality genre gap to recommender system design trade-offs between engagement optimization and satisfaction optimization.

---
*Always stop the Spark session when the notebook is finished to free JVM memory.*

In [ ]:
spark.stop()